### PR TITLE
#gh10501 extended payment endpoint to support allocation lines

### PR DESCRIPTION
**Pull request by adi-stefan**
https://github.com/metasfresh/metasfresh/issues/10501 
There are a lot of changes files due to moving the SwaggerDocConstants classs into misc/de-metas-common.

### CODE REVIEW

File: backend/de.metas.business/src/main/java/de/metas/invoice/service/impl/AbstractInvoiceDAO.java [hunk 1]
```diff
@@ -413,4 +420,61 @@ public boolean hasCompletedInvoicesReferencing(@NonNull final InvoiceId invoiceI
...
+			return getInvoiceIdByDocumentIdIfExists(query);
+		}
+		return Optional.empty();
+	}
+
+	private Optional<InvoiceId> getInvoiceIdByDocumentIdIfExists(@NonNull final InvoiceQuery query)
+	{
+		final String documentNo = assumeNotNull(query.getDocumentNo(), "Param query needs to have a non-null docId; query={}", query);
+		final OrgId orgId = assumeNotNull(query.getOrgId(), "Param query needs to have a non-null orgId; query={}", query);
+		final DocBaseAndSubType docType = assumeNotNull(query.getDocType(), "Param query needs to have a non-null docType; query={}", query);
```
Comment (metas-ts) [inline backend/de.metas.business/src/main/java/de/metas/invoice/service/impl/AbstractInvoiceDAO.java:445]: Same remarks as with the orderDAO (=>also use DocSubType and fail if more than one)

File: backend/de.metas.business/src/main/java/de/metas/invoice/service/impl/AbstractInvoiceDAO.java [hunk 2]
```diff
@@ -413,4 +420,61 @@ public boolean hasCompletedInvoicesReferencing(@NonNull final InvoiceId invoiceI
...
+	{
+		final ExternalId externalId = assumeNotNull(query.getExternalId(), "Param query needs to have a non-null externalId; query={}", query);
+		final OrgId orgId = assumeNotNull(query.getOrgId(), "Param query needs to have a non-null orgId; query={}", query);
+
+		final IQueryBuilder<I_C_Invoice> queryBuilder = createQueryBuilder(I_C_Invoice.class)
+				.addOnlyActiveRecordsFilter()
+				.addEqualsFilter(I_C_Invoice.COLUMNNAME_AD_Org_ID, orgId)
+				.addEqualsFilter(I_C_Invoice.COLUMNNAME_ExternalId, externalId.getValue());
+
+		final int invoiceRepoId = queryBuilder.create().firstId();
```
Comment (metas-ts) [inline backend/de.metas.business/src/main/java/de/metas/invoice/service/impl/AbstractInvoiceDAO.java:468]: fail if there is more than one ID

File: backend/de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/payment/JsonPaymentService.java [hunk 3]
```diff
@@ -144,4 +243,64 @@ private OrgId retrieveOrg(@RequestBody @NonNull final JsonInboundPaymentInfo jso
 		// TODO it would be nice here if we would also use orgID when searching for the bpartner, since an ExternalId may belong to multiple users from different orgs.
```
Comment (metas-ts) [inline backend/de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/payment/JsonPaymentService.java:243]: ...not just nice..TODO: 
* add OrgId as a parameter OR
* make sure it fails if there are multiple bpartners with the same externalId OR
* don't use bpartnerPriceListServicesFacade but something else (which then filters by org)

**Review by metas-ts — CHANGES_REQUESTED** [review]

File: backend/de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/payment/JsonPaymentService.java (hunk 3 above)
Comment (adi-stefan) [reply backend/de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/payment/JsonPaymentService.java]: added OrgId as param.

File: backend/de.metas.business/src/main/java/de/metas/invoice/service/impl/AbstractInvoiceDAO.java (hunk 2 above)
Comment (adi-stefan) [reply backend/de.metas.business/src/main/java/de/metas/invoice/service/impl/AbstractInvoiceDAO.java]: changed to firstIdOnly

File: backend/de.metas.business/src/main/java/de/metas/invoice/service/impl/AbstractInvoiceDAO.java (hunk 1 above)
Comment (adi-stefan) [reply backend/de.metas.business/src/main/java/de/metas/invoice/service/impl/AbstractInvoiceDAO.java:445]: added DocSubType and logic to use it.

File: misc/de-metas-common/de-metas-common-rest_api/src/main/java/de/metas/common/rest_api/payment/JsonInboundPaymentInfo.java [hunk 4]
```diff
@@ -55,15 +58,10 @@
...
-	@NonNull
-	String externalOrderId;
-
-	@ApiModelProperty(required = true)
-	@NonNull
-	BigDecimal amount;
+	@ApiModelProperty(dataType = "java.lang.String", //
+			value = SwaggerDocConstants.ORDER_IDENTIFIER_DOC)
+	@Nullable
+	String orderIdentifier;
```
[the remark this answers is not in the document]
Comment (adi-stefan) [reply misc/de-metas-common/de-metas-common-rest_api/src/main/java/de/metas/common/rest_api/payment/JsonInboundPaymentInfo.java:64]: Added JsonInboundPaymentInfoTest and JsonPaymentAllocationLineTest.

File: misc/de-metas-common/de-metas-common-rest_api/src/main/java/de/metas/common/rest_api/SwaggerDocConstants.java [hunk 5]
```diff
@@ -30,6 +30,14 @@
 			+ "* or something like `val-<C_Bartner.Value>`\n"
 			+ "* or something like `gln-<C_Bartner_Location.GLN>`\n";
 
+	public static final String INVOICE_IDENTIFIER_DOC = "Identifier of the Invoice in question. Can be\n"
+			+ "* a plain `<C_Invoice.C_Invoice_ID>`\n"
+			+ "* or something like `ext-<C_Invoice.ExternalId>`";
```
[the remark this answers is not in the document]
Comment (adi-stefan) [reply misc/de-metas-common/de-metas-common-rest_api/src/main/java/de/metas/common/rest_api/SwaggerDocConstants.java:36]: added

File: backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/order/OrderQuery.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+import lombok.Builder;
+import lombok.NonNull;
+import lombok.Value;
+
+import javax.annotation.Nullable;
+
+@Value
+public class OrderQuery
+{
+	Integer orderId;
```
[the remark this answers is not in the document]
Comment (adi-stefan) [reply backend/de.metas.adempiere.adempiere/base/src/main/java/de/metas/order/OrderQuery.java:37]: As discussed, OrderId not visible from this class.

File: backend/de.metas.business/src/main/java/de/metas/order/impl/AbstractOrderDAO.java [hunk 7]
```diff
@@ -328,4 +339,37 @@ public void save(@NonNull final org.compiere.model.I_C_OrderLine orderLine)
...
+	{
+		final String documentNo = assumeNotNull(query.getDocumentNo(), "Param query needs to have a non-null document number; query={}", query);
+		final OrgId orgId = assumeNotNull(query.getOrgId(), "Param query needs to have a non-null orgId; query={}", query);
+		final DocBaseAndSubType docType = assumeNotNull(query.getDocType(), "Param query needs to have a non-null docType; query={}", query);
+
+		final IQueryBuilder<I_C_Order> queryBuilder = createQueryBuilder()
+				.addOnlyActiveRecordsFilter()
+				.addEqualsFilter(I_C_Order.COLUMNNAME_AD_Org_ID, orgId)
+				.addEqualsFilter(I_C_Order.COLUMNNAME_DocumentNo, documentNo)
+				.addEqualsFilter(I_C_Order.COLUMNNAME_C_DocType_ID, NumberUtils.asInt(docType.getDocBaseType(), -1));
```
[the remark this answers is not in the document]
Comment (adi-stefan) [reply backend/de.metas.business/src/main/java/de/metas/order/impl/AbstractOrderDAO.java:370]: using docSubType.

File: backend/de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/payment/JsonPaymentService.java [hunk 8]
```diff
@@ -103,25 +133,94 @@ public JsonPaymentService(final CurrencyService currencyService, final BpartnerP
...
+						.dateTrx(dateTrx)
+						.createAndProcess();
 
+				final String orderIdentifier = jsonInboundPaymentInfo.getOrderIdentifier();
+				if (!Check.isEmpty(orderIdentifier))
+				{
+					Optional<String> externalOrderId = getExternalOrderIdFromIdentifier(IdentifierString.of(orderIdentifier), orgId);
+					if (!externalOrderId.isPresent())
+					{
+						logger.debug("could not find externalOrderId for identifier: " + orderIdentifier);
```
[the remark this answers is not in the document]
Comment (adi-stefan) [reply backend/de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/payment/JsonPaymentService.java]: Change to throw error via assert.

File: backend/de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/payment/JsonPaymentService.java [hunk 9]
```diff
@@ -103,25 +133,94 @@ public JsonPaymentService(final CurrencyService currencyService, final BpartnerP
...
+				.dateAcct(payment.getDateAcct())
+				.orgId(orgId);
+		for (JsonPaymentAllocationLine line : lines)
+		{
+			final String invoiceId = line.getInvoiceIdentifier();
+			final DocBaseAndSubType docType = DocBaseAndSubType.of(line.getDocType());
+			final Optional<InvoiceId> invoice = retrieveInvoice(IdentifierString.of(invoiceId), OrgId.ofRepoIdOrNull(orgId), docType);
+			if (!invoice.isPresent())
+			{
+				logger.warn("Cannot find invoice for identifier: " + invoiceId);
```
[the remark this answers is not in the document]
Comment (adi-stefan) [reply backend/de.metas.business.rest-api-impl/src/main/java/de/metas/rest_api/payment/JsonPaymentService.java]: changed to throw error via assert.

File: backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/invoicecandidate/api/impl/InvoiceCandBLCreateInvoices.java [hunk 10]
```diff
@@ -349,16 +349,21 @@ private I_C_Invoice createInvoice(
 				invoice.setDateAcct(TimeUtil.asTimestamp(invoiceHeader.getDateAcct(), timeZone)); // 03905: also updating DateAcct
 
 				invoice.setM_PriceList_ID(invoiceHeader.getM_PriceList_ID()); // #367: get M_PriceList_ID directly from invoiceHeader.
+				Set<String> externalIds = invoiceHeader.getAllInvoiceCandidates().stream().map(cand -> cand.getExternalHeaderId()).filter(cand -> cand != null).collect(Collectors.toSet());
+				if (externalIds.size() > 1)
+				{
+					logger.warn("Unexpectedly found multiple externalId candidates for the same invoice: " + externalIds.toArray());
```
[the remark this answers is not in the document]
Comment (adi-stefan) [reply backend/de.metas.swat/de.metas.swat.base/src/main/java/de/metas/invoicecandidate/api/impl/InvoiceCandBLCreateInvoices.java]: changed to an assert, but also added externalId to IInvoiceHeader/InvoiceHeaderImpl/InvoiceHeaderImplBuilder  logic.

File: misc/de-metas-common/de-metas-common-rest_api/src/main/java/de/metas/common/rest_api/payment/JsonPaymentAllocationLine.java [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+{
+	@NonNull
+	@ApiModelProperty(required = true,
+			dataType = "java.lang.String",
+			value = SwaggerDocConstants.INVOICE_IDENTIFIER_DOC)
+	String invoiceIdentifier;
+
+	@ApiModelProperty(position = 10)
+	@Nullable
+	String docType;
```
[the remark this answers is not in the document]
Comment (adi-stefan) [reply misc/de-metas-common/de-metas-common-rest_api/src/main/java/de/metas/common/rest_api/payment/JsonPaymentAllocationLine.java]: renamed docType -> docBaseType and added docSubType

File: misc/de-metas-common/de-metas-common-rest_api/src/main/java/de/metas/common/rest_api/payment/JsonInboundPaymentInfo.java [hunk 12]
```diff
@@ -47,6 +47,12 @@
 @JsonDeserialize(builder = JsonInboundPaymentInfo.JsonInboundPaymentInfoBuilder.class)
 public class JsonInboundPaymentInfo
 {
+	@ApiModelProperty(required = true, //
+			dataType = "java.lang.String", //
+			value = "An identifier for the transaction being processed.")
+	@Nullable
+	String paymentId;
```
Comment (metas-ts) [inline misc/de-metas-common/de-metas-common-rest_api/src/main/java/de/metas/common/rest_api/payment/JsonInboundPaymentInfo.java:54]:
```suggestion
			value = "An external identifier for the payment being posted to metasfresh. Translates to `C_Payment.ExternalId`")
	@Nullable
	String externalPaymentId;
```

...to make sure noone thinks it's about `C_Payment.C_Payment_ID`..

**Review by metas-ts — CHANGES_REQUESTED** [review]
..just some tiny requests

**Review by metas-ts — APPROVED** [review]

**Pull request merged**